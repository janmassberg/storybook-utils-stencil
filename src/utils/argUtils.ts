export type ArgEntry = [string, any];
export type ArgsType = ArgEntry[];

export interface ArgsFilter {
    (args: any): ArgsType;
}

/** @private */
const isFunctionValue = (value: any): boolean => {
    return typeof value === "function";
};

/** @private */
const isObjectValue = (value: any): boolean => {
    return typeof value === "object" && value !== null;
};

/** @private */
const isEventHandlerKey = (key: string): boolean => {
    return /^on[A-Z][a-zA-Z]+$/.test(key);
};

/** @private */
const isHiddenArg = ([key]: ArgEntry): boolean => {
    return key.slice(0, 1) === "_";
};

/**
 * Checks if an argument is an event handler
 * @param args {ArgsType}
 */
export const isEventArg = ([key, value]: ArgEntry): boolean => {
    return isEventHandlerKey(key) && isFunctionValue(value);
};

/**
 * Checks if an argument is a valid HTML attribute
 * @param args {ArgsType}
 */
export const isHtmlArg = ([key, value]: ArgEntry): boolean => {
    return (
        value !== undefined &&
        value !== null &&
        !isHiddenArg([key, value]) &&
        !isEventArg([key, value]) &&
        !isFunctionValue(value) &&
        !isObjectValue(value)
    );
};

/**
 * Checks if an argument is a valid JSX property
 * @param args {ArgsType}
 */
export const isJsxArg = ([key, value]: ArgEntry): boolean => {
    return (
        value !== undefined &&
        value !== null &&
        !isHiddenArg([key, value]) &&
        !isEventArg([key, value]) &&
        !isHtmlArg([key, value])
    );
};

/**
 * Filters args with valid HTML attributes only
 * @param args {ArgsType}
 */
export const filterHtmlAttributes: ArgsFilter = (args: ArgsType): ArgsType => {
    return Object.entries(args).filter(isHtmlArg);
};

/**
 * Filters args with valid JSX properties only
 * @param args {ArgsType}
 */
export const filterJsxProperties: ArgsFilter = (args: ArgsType): ArgsType => {
    return Object.entries(args).filter(isJsxArg);
};

/**
 * Filters args with valid event handlers only
 * @param args {ArgsType}
 */
export const filterEventHandlers: ArgsFilter = (args: ArgsType): ArgsType => {
    return Object.entries(args).filter(isEventArg);
};
