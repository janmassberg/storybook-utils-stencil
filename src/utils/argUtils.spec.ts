import { filterHtmlAttributes, filterJsxProperties } from "./argUtils";

const testArgs = {
    testString: "test-string",
    testNumber: 12345,
    testTrue: true,
    testFalse: false,
    testArray: [1, 2, 3],
    testObject: {
        foo: "bar",
    },
    testFunction: () => {
        // do nothing
    },
    testNull: null,
    testUndefined: undefined,
};

describe("filterHtmlAttributes", () => {
    it("should return an array of key/value pairs with valid HTML attributes", () => {
        expect(filterHtmlAttributes(testArgs)).toMatchSnapshot();
    });
});

describe("filterJsxProperties", () => {
    it("should return an array of key/value pairs with valid JSX properties", () => {
        expect(filterJsxProperties(testArgs)).toMatchSnapshot();
    });
});
