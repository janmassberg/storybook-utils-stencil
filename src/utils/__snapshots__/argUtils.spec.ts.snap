// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filterHtmlAttributes should return an array of key/value pairs with valid HTML attributes 1`] = `
Array [
  Array [
    "testString",
    "test-string",
  ],
  Array [
    "testNumber",
    12345,
  ],
  Array [
    "testTrue",
    true,
  ],
  Array [
    "testFalse",
    false,
  ],
  Array [
    "testNull",
    null,
  ],
]
`;

exports[`filterJsxProperties should return an array of key/value pairs with valid JSX properties 1`] = `
Array [
  Array [
    "testArrayt",
    Array [
      1,
      2,
      3,
    ],
  ],
  Array [
    "testObject",
    Object {
      "foo": "bar",
    },
  ],
  Array [
    "testFunction",
    [Function],
  ],
]
`;
